odoo/addons/base/models/ir_qweb.py

```diff
@@ -148,31 +722,880 @@ def is_child_view(view_name):
+        # * similarly for a comprehensions [a + b for a in c] should be
+        #     compiledto [a + values.get('b') for a in values.get('c')]
+        # to avoid the risk of confusion between nested lambdas / comprehensions,
+        # this is currently performed independently at each level of brackets
+        # nesting (hence the function being recursive).
+        index = 0
+        open_bracket_index = -1
+        bracket_depth = 0
+
+        argument_name = '_arg_%s__'
+        argument_names = argument_names or []
+
+        while index < len(tokens):
+            t = tokens[index]
+            if t.exact_type in [token.LPAR, token.LSQB, token.LBRACE]:
+                bracket_depth += 1
+            if t.exact_type in [token.RPAR, token.RSQB, token.RBRACE]:
+                bracket_depth -= 1
+            elif bracket_depth == 0 and t.exact_type == token.NAME:
+                string = t.string
+                if string == 'lambda': # lambda => allowed values for the current bracket depth
+                    i = index + 1
+                    while i < len(tokens):
+                        t = tokens[i]
+                        if t.exact_type == token.NAME:
+                            argument_names.append(t.string)
+                        elif t.exact_type == token.COMMA:
+                            pass
+                        elif t.exact_type == token.COLON:
+                            break
+                        elif t.exact_type == token.EQUAL:
+                            raise NotImplementedError('Lambda default values are not supported')
+                        else:
+                            raise NotImplementedError('This lambda code style is not implemented.')
+                        i += 1
+                elif string == 'for': # list comprehensions => allowed values for the current bracket depth
+                    i = index + 1
+                    while len(tokens) > i:
+                        t = tokens[i]
+                        if t.exact_type == token.NAME:
+                            if t.string == 'in':
+                                break
+                            argument_names.append(t.string)
+                        elif t.exact_type in [token.COMMA, token.LPAR, token.RPAR]:
+                            pass
+                        else:
+                            raise NotImplementedError('This loop code style is not implemented.')
+                        i += 1
+
+            index += 1
+
+        # Use bracket to nest structures.
+        # Recursively processes the "sub-scopes", and replace their content with
+        # a compiled node. During this recursive call we add to the allowed
+        # values the values provided by the list comprehension, lambda, etc.,
+        # previously extracted.
+        index = 0
+        open_bracket_index = -1
+        bracket_depth = 0
+
+        while index < len(tokens):
```

Comment:
It works, for the rest if you have an idea to rewrite it it can be done later but here I have not modified it (it had fixes / improvement by stéphane, I do not want to interject regression with another implementation)